### ISSUE TITLE
how to save model(nni-free) after quantization and pruners? 

### ISSUE THREAD

**Issue by Greyman-Seu** [microsoft/nni#2097]
**Short summary about the issue/question**:

**Brief what process you are following**: 
ModelSpeedup just shorten the runtime? when saving the model(torch.save) after quantization and pruners,the model is getting bigger.Do you have the plan to solve the problem? 

**How to reproduce it**: 

**nni Environment**:
- nni version:1.4
- nni mode(local|pai|remote):
- OS:window/liunx
- python version:3.6
- is conda or virtualenv used?: conda
- is running in docker?:no

**need to update document(yes/no)**:

**Anything else we need to know**:

**Comment by QuanluZhang** [comment]
@Greyman-Seu Currently, ModelSpeedup in NNI is only for speeding up the model pruned by coarse-grained pruning algorithms. The speedup of quantization has not been supported yet.

For exporting the faster model after ModelSpeedup, you could try to export it as an onnx model.

**Comment by Greyman-Seu** [comment]
> @Greyman-Seu Currently, ModelSpeedup in NNI is only for speeding up the model pruned by coarse-grained pruning algorithms. The speedup of quantization has not been supported yet.
> 
> For exporting the faster model after ModelSpeedup, you could try to export it as an onnx model.

thanks for reply~  Can I think that 1) any prunning algorithms make the model sparse,because the 
prune model file(pth->zip) becomes smaller after compression than the raw model(pth->zip). The sparse model(.pth) can be used to short runtime in some specific hardware. 2)the pruned model *before* ModelSpeedup could be export as an onnx model, is the onnx model sparse? 3）currently,My model(.pth) can‘t be smaller after quantization? when the quantized model  export as an onnx model，could this model(.onnx)  be smaller and be used in pytorch framework?

**Comment by Cjkkkk** [comment]
@Greyman-Seu Currently, the type of tensor after quantization is still original type, therefore the size is not smaller.

**Comment by Greyman-Seu** [comment]
> @Greyman-Seu Currently, the type of tensor after quantization is still original type, therefore the size is not smaller.
I use torch.save() . The size of quantized model is bigger. @QuanluZhang @Cjkkkk

**Comment by QuanluZhang** [comment]
@Greyman-Seu A model is called sparse if there are many zeros in its weights. but this does not mean the model is faster or the model is smaller. Because the computation still exists even the weights are zero.

Usually, compressing a model can be split into two steps: 1 finding the best masks for weights, 2 speed up the model based on the masks. The pruning algorithms do step 1, and the ModelSpeedup feature does step 2.

**Comment by Greyman-Seu** [comment]
> @Greyman-Seu A model is called sparse if there are many zeros in its weights. but this does not mean the model is faster or the model is smaller. Because the computation still exists even the weights are zero.
> 
> Usually, compressing a model can be split into two steps: 1 finding the best masks for weights, 2 speed up the model based on the masks. The pruning algorithms do step 1, and the ModelSpeedup feature does step 2.

I am new at model compression. The pruned model before Modelspeedup is sparse because there are many zeros in its weights. This model could be speed up by some engine of acceleration just like SparseBLAS(tensorflow lite) even this engine is is not particularly good .   

how to save an onnx model after modelspeedup? could you give an example including the step 1, step2 and export the faster model(.onnx)? 
(model_speedup.py)
m_speedup cant export an onnx model. Using torch.save(model,"model.pth"), the size of model is same as raw model.

**Comment by QuanluZhang** [comment]
step1 run examples/model_compress/slim_torch_cifar10.py

step2 run model_speedup.py

at last, for exporting the model, adding one line `torch.onnx.export(model, dummy_input.to(device), 'after_speedup.onnx')` in model_speedup.py

**Comment by Greyman-Seu** [comment]
> step1 run examples/model_compress/slim_torch_cifar10.py
> 
> step2 run examples/model_compress/model_speedup.py
> 
> at last, for exporting the model, adding one line `torch.onnx.export(model, dummy_input.to(device), 'after_speedup.onnx')` in model_speedup.py

@QuanluZhang   
 when running model_speedup.py after step 1(slim_torch_cifar10.py), get error massage like "KeyError:"feature.1"". (nni-1.4.0   torch-1.4.0   torchvision-0.5.0)

==>self.name_to_gnode[module_name].op_type
self.name_to_gnode is empty dictionary.

**Comment by QuanluZhang** [comment]
@Greyman-Seu ModelSpeedup in Alpha version only supports torch 1.3.1 and torchvision 0.4.2, torch 1.4 will be supported in v1.5 release. You could refer to [ModelSpeedup doc](https://nni.readthedocs.io/en/latest/Compressor/ModelSpeedup.html) for details.

**Comment by Greyman-Seu** [comment]
thanks! @QuanluZhang   I have another question.After running "DoReFaQuantizer_torch_mnist.py"，It seems to actually increase the size of the model (from ~1.64Mb to ~3.29Mb). I understand that nni does not support real quantization. Do you have an plan to solve it???

**Comment by QuanluZhang** [comment]
@Greyman-Seu according to quantization algorithms, the weights before quantize are also kept and trained. This is for getting high performance model. If you quantize every layer in the model, the model size would be doubled. We will add an `export_model` method for `Quantizer` just like that in `Pruner`.

**Comment by scarlett2018** [comment]
Closing the issue as the question has been answered, thanks @Greyman-Seu and @QuanluZhang
